### ISSUE TITLE
vm_armv7l.c is not used by default

### ISSUE THREAD

**Issue by zturtleman** [ioquake/ioq3#291]
vm_armv7l.c should be used by default on armv7l but it is not. It's possible to force using it by running `make ARCH=armv7l` though.

The Makefile was changed to generalize armv6l, armv7l etc architectures as "arm".[[1]](https://github.com/ioquake/ioq3/commit/c91fab3900345d047f085d1e467521bba1ebd28f) Using vm_armv7l.c requires ARCH Makefile variable to be armv7l.[[2]](https://github.com/ioquake/ioq3/pull/91/files)

**Comment by qwertyuiop924** [comment]
I attempted a fix for this here: https://github.com/qwertyuiop924/ioq3/commit/da6c323b81782048d53fde91b5d509de70d3c964

However, I'm not willing to submit a PR until someone's looked it over: I don't have an ARM toolchain available at present, so I can't actually test it.

Noted issues: At present, I am not sure if `ARM_VERSION` needs to be `export`ed. It doesn't seem to be? but if it does, that can be done. Also, this doesn't really handle cross-compilation... at all. If we're cross compiling, you have to set `ARM_VERSION` by hand. I think this is The Right Thing, because we should assume lowest common denominator for the arch unless the user tells us otherwise, but that means ARM_VERSION needs to be documented, and I don't know where that documentation should be.

My apologies if I've violated style or something, this is my first attempt at contributing.

**Comment by zturtleman** [comment]
I think it would simpler to change COMPILE_ARCH to not change armv7l to arm. Then it will automatically enable armv7l VM JIT if supported for current system and cross-compiling can use `make ARCH=arm` or `make ARCH=armv7l` (with other `CC` etc variables for arm compiler).

It seems like ideally the armv7l VM JIT would always be compiled and check if CPU supports armv7l instructions at run-time. So there is only a single executable to distribute and `ARCH` can always be "arm". Though I don't know enough about it to say if that's possible. (For example, the Raspberry Pi 1 is armv6 and Raspberry Pi 2 and 3 are armv7.)

**Comment by ensiform** [comment]
Most don't distribute mods in bin form, because they aren't pure capable in q3. Only a few do (for sever?) But client and ui are usually qvm which only needs interpreter.

**Comment by qwertyuiop924** [comment]
This is true! However, there is an exception: *us*. We ship our gamecode as native code. 

Beyond that, even if most people don't ship native code, it's still a feature we advertise, and a feature we provide on every platform. Breaking native code modules (or more accurately, making them needlessly difficult to load) on one platform is at best a Bad Idea and at worst kind of insane.

**Comment by ensiform** [comment]
I'm not familiar with what mod you are part of.

**Comment by qwertyuiop924** [comment]
Sorry. By "we" I meant ioquake3. Our nightlies and general builds contain native gamecode.

**Comment by ensiform** [comment]
They do, but they cannot be used on servers.

**Comment by ensiform** [comment]
Native code support is for development only, especially for the nonstandard formats. the original game didn't support arm, so unless there are now actual mods being used with the old style filename, it's not really an issue. You could support both if need be.

**Comment by zturtleman** [comment]
> By the looks if it, it's certainly possible. The only problem is portability. Oh, and adding a big fat #ifdef __arm__ (or your regional equivalent: probably actually setting arm bits in q_platform and checking those) right in the middle of VM_Create, which may offend taste, because... well, that's pretty gross, innit. I guess technically you don't have to, but then we'd be running arm code on every system, which is dumb.

I meant compiling vm_armv7l.c on all `arm*` architectures and make vm_armv7l.c's VM_Compile() check if it can be used. If it cannot, vm_armv7l.c's VM_Compile() should set `vm->compiled = qfalse` and return. This causes VM_Create() to fallback to byte code interpreter. No changes needed outside of vm_armv7l.c.

https://github.com/ioquake/ioq3/blob/e5da13f14635b6accc5eefd6986de03d2cff9662/code/qcommon/vm.c#L654-L672

**Comment by qwertyuiop924** [comment]
>I meant compiling vm_armv7l.c on all arm* architectures and make vm_armv7l.c's VM_Compile() check if it can be used. 

D'oh. I don't know how I missed that. Sorry.

I'd be willing to try my hand at implementing this, although I confess to being new to the codebase, so... yeah. I'll try not to mess things up.

>the original game didn't support arm, so unless there are now actual mods being used with the old style filename, it's not really an issue. You could support both if need be.

The issue is that with the proposed change, there wouldn't be a new style: the filename required would actually change based on compilation conditions.

**Comment by qwertyuiop924** [comment]
Alright, here's my attempt at adding a runtime check and doing the fix that way: https://github.com/qwertyuiop924/ioq3/commit/501c9f66947c7f2966a7dac937190a9d7b1325f9

I'm not sure if the code is up to standards, and I'd be happy to fix it if it isn't.

The bigger problem, however, is that I don't have an arm development environment at present. So I can't test this.

**Comment by abcd1234-byte** [comment]
It runs right on my Raspberry pi except the fact it lags a bit, but that can be fixed with the opengl1 renderer enabled.

**Comment by abcd1234-byte** [comment]
just use q3lite

**Comment by NuclearMonster** [comment]
> just use q3lite
I understand it might be frustrating but please understand it isn't useful feedback for this project.

**Comment by abcd1234-byte** [comment]
maybe you should integrate github.co/cdev-tux/q3lite renderer and vm into mainline.

**Comment by zturtleman** [comment]
q3lite relicensed as GPLv3 with additional terms from Zenimax. As is, it cannot be merged into ioquake3, A polished OpenGL ES 1 renderer is sitting in #375 though.
